Bucket in path

hultin opened the pull request.

Commit:
Formatting with black

Commit:
WIP, works with bucket but not with schema

Commit:
Reworked settings

Commit:
Multiple boto clients per schema

Commit:
Removing AWS_S3_BUCKET_NAME and AWS_S3_ENDPOINT_URL

Commit:
Remove sync_meta and the managment command

Commit:
Remove StaticS3Storage and ManifestStaticS3Storage

Commit:
Removing AWS_S3_BUCKET_AUTH and AWS_S3_PUBLIC_URL

Commit:
v1928.15.dev0

Commit:
Re-add AWS_S3_BUCKET_NAME to be compatible with old migrations.

Old migration should be squashed and after that, we could remove AWS_S3_BUCKET_NAME again.

Commit:
_read_only_protect

Commit:
validate_s3_path

Commit:
Small suggestions on code clean up and formatting

oskarerikmartinandersson commented on tests/django_s3_storage_test/settings.py:
```diff
@@ -16,28 +16,17 @@
 
-STATICFILES_STORAGE = "django_s3_storage.storage.ManifestStaticS3Storage"
-
+DEBUG = True
```
What implications does this have?

hultin replied:
It is useful during debugging (duh!) Django. More details on error pages and better traces on exceptions. But no impact on django-s3-storage itself. 

I don't think we needed anymore for now, so I will remove it.

oskarerikmartinandersson commented:
How do think about documentation in this repository? Should we updated it to match the code changes, or should we remove it so it's not out-of-sync with the code?

We also have other package/project related stuff we should maybe think about, e.g.
- Changelog
- License 
- Package related code (e.g. supported versions)
- Github related code

oskarerikmartinandersson commented on tests/django_s3_storage_test/settings.py:
```diff
@@ -46,6 +35,6 @@
     "default": {
         "ENGINE": "django.db.backends.sqlite3",
-        "NAME": os.path.join(BASE_DIR, "db.sqlite3"),
+        "NAME": 'mydatabase',
```
Is there a reason to change?

hultin replied:
Nope, I was for debugging. Reverted.

oskarerikmartinandersson commented on django_s3_storage/storage.py:
I'm not sure on what order is best for classes, constants, functions etc. But right now everything seems a bit mixed, maybe move `Endpoints` down to the other classes?

oskarerikmartinandersson commented on django_s3_storage/storage.py:
```diff
@@ -92,34 +123,81 @@ def open(self, mode="rb"):
+    AWS_S3_MAX_POOL_CONNECTIONS: int = 10
+    AWS_S3_CONNECT_TIMEOUT: int = 60  # 60 seconds
+    AWS_S3_BUCKET_NAME: str = 'Deprecated'
+    AWS_S3_READ_ONLY: bool = False
```
Keep order from original settings, i.e. move `AWS_S3_BUCKET_NAME` back up

```suggestion
    # default_s3_settings
    AWS_S3_BUCKET_NAME: str = 'Deprecated'
    AWS_S3_ADDRESSING_STYLE: str = "auto"
    AWS_S3_ENDPOINTS: dict = field(
        default_factory=lambda: {
            's3': Endpoints(),
        },
    )
    AWS_S3_KEY_PREFIX: str = ""
    AWS_S3_MAX_AGE_SECONDS: int = 60 * 60  # 1 hours.
    AWS_S3_REDUCED_REDUNDANCY: bool = False
    AWS_S3_CONTENT_DISPOSITION: str = ""
    AWS_S3_CONTENT_LANGUAGE: str = ""
    AWS_S3_METADATA: dict = field(default_factory=dict)
    AWS_S3_ENCRYPT_KEY: bool = False
    AWS_S3_KMS_ENCRYPTION_KEY_ID: str = ""
    AWS_S3_GZIP: bool = True
    AWS_S3_SIGNATURE_VERSION: str = "s3v4"
    AWS_S3_FILE_OVERWRITE: bool = False
    AWS_S3_USE_THREADS: bool = True
    AWS_S3_MAX_POOL_CONNECTIONS: int = 10
    AWS_S3_CONNECT_TIMEOUT: int = 60  # 60 seconds
    AWS_S3_READ_ONLY: bool = False
```

oskarerikmartinandersson commented on django_s3_storage/storage.py:
```diff
@@ -92,34 +123,81 @@ def open(self, mode="rb"):
 
-    """
-    Thread-local connection manager.
+    def __post_init__(self):
```
Do we this method?

hultin replied:
Leftover. Removed

oskarerikmartinandersson commented on django_s3_storage/storage.py:
```diff
@@ -129,137 +207,103 @@ class S3Storage(Storage):
-        "AWS_S3_CONNECT_TIMEOUT": 60  # 60 seconds
-    }
-
     s3_settings_suffix = ""
```
Should this attribute be removed? I can't find any reference to it anymore?

hultin replied:
Yes, leftover from StaticS3Storage and ManifestStaticS3Storage. Removed.

oskarerikmartinandersson commented on django_s3_storage/storage.py:
```diff
@@ -369,40 +426,64 @@ def get_available_name(self, name, max_length=None):
+        ok_bucket = url_split.netloc != ''
+        if not ok_schema or not ok_bucket:
+            raise NotAS3Url(
+                f"The filename {s3_path} is not a full S3 URL. Have you forgotten to set 'upload_to' on the FileField?"
```
Is this error message a 1928 specific implementation or is this general to the overall changes you have made?

hultin replied:
It is to the general changes.

oskarerikmartinandersson commented on django_s3_storage/storage.py:
```diff
@@ -446,24 +529,15 @@ def size(self, name):
-            if extra_params or client_method != "get_object":
-                raise ValueError("Use of extra_params or client_method is not allowed with AWS_S3_PUBLIC_URL")
-            return urljoin(self.settings.AWS_S3_PUBLIC_URL, filepath_to_uri(name))
         # Otherwise, generate the URL.
```
Comment doesn't work after above code was removed

oskarerikmartinandersson commented on tests/django_s3_storage_test/tests.py:
```diff
@@ -15,20 +14,29 @@
-        for entry in default_storage.listdir(""):
-            default_storage.delete("/".join(entry))
+class TestS3Storage(SimpleTestCase):
+    # def tearDown(self):
+    #     # clean up the dir
+    #     for entry in default_storage.listdir(helper_old_to_new_name("")):
+    #         print(entry)
+    #         # default_storage.delete("/".join(entry))
```
Remove unused code.

```suggestion
```

oskarerikmartinandersson commented on tests/django_s3_storage_test/tests.py:
```diff
@@ -15,20 +14,29 @@
-    def save_file(self, name="foo.txt", content=b"foo", storage=default_storage):
+    def save_file(
+        self,
+        name=f"s3://1928-django-s3-storage-test/foo.txt",
```
Unnecessary f-string

```suggestion
        name="s3://1928-django-s3-storage-test/foo.txt",
```

oskarerikmartinandersson commented on tests/django_s3_storage_test/tests.py:
Unused imports, should be:
```python
import time
from contextlib import contextmanager
from datetime import timedelta
from urllib.parse import urlsplit, urlunsplit

import requests
from django.core.exceptions import ImproperlyConfigured
from django.core.files.base import ContentFile
from django.core.files.storage import default_storage
from django.test import SimpleTestCase
from django.utils import timezone
from django.utils.timezone import is_naive, make_naive, utc
```

hultin committed:
Update tests/django_s3_storage_test/tests.py

Co-authored-by: oskarerikmartinandersson <37584173+oskarerikmartinandersson@users.noreply.github.com>

hultin committed:
Update django_s3_storage/storage.py

Keep order from original settings, i.e. move AWS_S3_BUCKET_NAME back up

Co-authored-by: oskarerikmartinandersson <37584173+oskarerikmartinandersson@users.noreply.github.com>

hultin committed:
Update tests/django_s3_storage_test/tests.py

Remove unused code.

Co-authored-by: oskarerikmartinandersson <37584173+oskarerikmartinandersson@users.noreply.github.com>

Commit:
Fix imports

Commit:
Update Readme.rst. Adding and removing configures.

Commit:
Update Readme.rst.

Short description of the fork and
removing paragraphs about features that are not supported anymore

Commit:
Read-only mode for presigned URL generation.

Commit:
Remove github actions.

Commit:
Reverting settings used for debugging during development.

Commit:
Add pytest.ini

Commit:
Add message to RuntimeError

Commit:
v1928.15.dev2

Commit:
Remove old comments that doesn't make sense after refactoring

oskarerikmartinandersson reviewed (approved):
Looks good to me, one question though have we lost some commit on this branch compared to the `v1928.15.dev2` tag? Should we re-apply the message in the `RuntimeError`?

![image](https://github.com/1928d/django-s3-storage/assets/37584173/d09057f8-59c2-41aa-9529-f3351e55d955)
